File: app/src/main/java/com/tuwien/buildinginteractioninterfaces/prototype/domain/model/BenchmarkModel.java
package com.tuwien.buildinginteractioninterfaces.prototype.domain.model;

import java.io.Serializable;
import java.util.Calendar;
import java.util.Date;

public class BenchmarkModel implements Serializable{
    private int correctChars;
    private int correctWords;
    private int errors;
    private float charsPerSec;
    private float wordsPerSec;
    private int totalWords;
    private float time;
    private int backspace;
    private int keystrokes;
    private int characters;
    private double keystrokesPerChar;
    private double minimumStringDistanceErrorRate;
    private OptionsModel options;
    private Date timestamp = Calendar.getInstance().getTime();

    public BenchmarkModel(OptionsModel options) {
        this.options = options;
    }

    public int getCorrectChars() {
        return correctChars;
    }

    public void setCorrectChars(int correctChars) {
        this.correctChars = correctChars;
    }

    public int getCorrectWords() {
        return correctWords;
    }

    public void setCorrectWords(int correctWords) {
        this.correctWords = correctWords;
    }

    public int getErrors() {
        return errors;
    }

    public void setErrors(int errors) {
        this.errors = errors;
    }

    public float getCharsPerSec() {
        return charsPerSec;
    }

    public void setCharsPerSec(float charsPerSec) {
        this.charsPerSec = charsPerSec;
    }

    public void incrementCorrectChars(){
        correctChars++;
    }
    public void incrementCorrectChars(int numChars){
        correctChars += numChars;
    }
    public void incrementCorrectWords(){
        correctWords++;
    }
    public void incrementErrors(){
        errors++;
    }

    public int getTotalWords() {
        return totalWords;
    }

    public void setTotalWords(int totalWords) {
        this.totalWords = totalWords;
    }

    public float getTime() {
        return time;
    }

    public void setTime(float time) {
        this.time = time;
    }

    public int getBackspace() {
        return backspace;
    }

    public void setBackspace(int backspace) {
        this.backspace = backspace;
    }

    public int getKeystrokes() {
        return keystrokes;
    }

    public void setKeystrokes(int keystrokes) {
        this.keystrokes = keystrokes;
        updateKeystrokesPerChar();
    }

    public int getCharacters() {
        return characters;
    }

    public void setCharacters(int characters) {
        this.characters = characters;
        updateKeystrokesPerChar();
    }

    public double getKeystrokesPerChar() {
        return keystrokesPerChar;
    }

    public void setKeystrokesPerChar(double keystrokesPerChar) {
        this.keystrokesPerChar = keystrokesPerChar;
    }

    public void updateKeystrokesPerChar(){
        keystrokesPerChar = (double) keystrokes / (double) characters;
    }

    public double getMinimumStringDistanceErrorRate(){
        return this.minimumStringDistanceErrorRate;
    }

    public void setMinimumStringDistanceErrorRate(double minimumStringDistanceErrorRate){
        this.minimumStringDistanceErrorRate = minimumStringDistanceErrorRate;
    }

    public OptionsModel getOptions() {
        return options;
    }

    public void setOptions(OptionsModel options) {
        this.options = options;
    }

    public float getWordsPerSec() {
        return wordsPerSec;
    }

    public void setWordsPerSec(float wordsPerSec) {
        this.wordsPerSec = wordsPerSec;
    }

    public Date getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }

    @Override
    public String toString() {
        return "BenchmarkModel{" +
                "correctChars=" + correctChars + "\n" +
                ", correctWords=" + correctWords + "\n" +
                ", errors=" + errors + "\n" +
                ", charsPerSec=" + charsPerSec + "\n" +
                ", wordsPerSec=" + wordsPerSec + "\n" +
                ", totalWords=" + totalWords + "\n" +
                ", time=" + time + "\n" +
                ", backspace=" + backspace + "\n" +
                ", keystrokes=" + keystrokes + "\n" +
                ", characters=" + characters + "\n" +
                ", keystrokesPerChar=" + keystrokesPerChar + "\n" +
                ", minimumStringDistanceErrorRate=" + minimumStringDistanceErrorRate + "\n" +
                ", options=" + options + "\n" +
                ", timestamp=" + timestamp + "\n" +
                '}';
    }
}

File: app/src/main/java/com/tuwien/buildinginteractioninterfaces/prototype/domain/model/BenchmarkModel.kt
package com.tuwien.buildinginteractioninterfaces.prototype.domain.model

import android.arch.persistence.room.ColumnInfo
import android.arch.persistence.room.Entity
import android.arch.persistence.room.PrimaryKey
import com.tuwien.buildinginteractioninterfaces.prototype.util.Benchmarks
import java.io.Serializable
import java.util.*

@Entity(tableName = "benchmark")
class BenchmarkModel(var options: OptionsModel, var keyboardApp: String): Serializable {
    @PrimaryKey(autoGenerate = true)
    var uid: Int = 0

    @ColumnInfo(name = "correct_chars")
    var correctChars: Int = 0
        set(value) {
            field = value
            updateKeystrokesPerChar()
            updateCharsPerSecond()
        }

    @ColumnInfo(name = "correct_words")
    var correctWords: Int = 0
        set(value) {
            field = value
            updateWordsPerMinute()
            updateWordsPerSec()
        }
    @ColumnInfo(name = "errors")
    var errors: Int = 0

    @ColumnInfo(name = "total_words")
    var totalWords: Int = 0

    @ColumnInfo(name = "time_elapsed")
    var timeElapsed:Long = 0 // In miliseconds
        set(value) {
            field = value
            updateWordsPerMinute()
            updateKeystrokesPerSecond()
            updateCharsPerSecond()
            updateWordsPerSec()
        }

    @ColumnInfo(name = "backspace")
    var backspace: Int = 0

    @ColumnInfo(name = "keystrokes")
    var keystrokes: Int = 0
        set(value) {
            field = value
            updateKeystrokesPerSecond()
            updateKeystrokesPerChar()
        }
    @ColumnInfo(name = "characters")
    var characters: Int = 0

    @ColumnInfo(name = "timestamp")
    var timestamp: Date = Calendar.getInstance().getTime()

    @ColumnInfo(name = "input_stream")
    var inputStream: StringBuffer = StringBuffer()

    @ColumnInfo(name = "transcribed_string")
    var transcribedString: StringBuffer = StringBuffer()

    //Entry rates
    @ColumnInfo(name = "words_per_minute")
    var wordsPerMinute: Float = 0.0f

    @ColumnInfo(name = "keystrokes_per_second")
    var keystrokesPerSecond: Float = 0.0f

    //Error rates
    @ColumnInfo(name = "keystrokes_per_char")
    var keystrokesPerChar: Double = 0.0

    @ColumnInfo(name = "minimum_string_distance_error_rate")
    var minimumStringDistanceErrorRate: Double = 0.0

    //Custom entry rates
    @ColumnInfo(name = "chars_per_sec")
    var charsPerSec: Float = 0.0f

    @ColumnInfo(name = "words_per_sec")
    var wordsPerSec: Float = 0.0f

    fun updateKeystrokesPerChar(){
        keystrokesPerChar = Benchmarks.keystrokesPerChar(keystrokes,correctChars)
    }

    fun updateWordsPerMinute(){
        wordsPerMinute = correctWords.toFloat()/ timeElapsed * 60 * 1000
    }

    fun updateKeystrokesPerSecond(){
        keystrokesPerSecond = (keystrokes.toFloat() - 1) / timeElapsed * 1000
    }

    fun updateCharsPerSecond(){
        charsPerSec = if(timeElapsed == 0L) 0.0f else correctChars.toFloat() / timeElapsed * 1000
    }

    fun updateWordsPerSec(){
        wordsPerSec = if(timeElapsed == 0L) 0.0f else correctWords.toFloat() / timeElapsed * 1000
    }

    fun addToTranscribedString(word: String){
        transcribedString.append(word + "\n")
    }

    fun addToInputStream(word: String){
        inputStream.append(word + "\n")
    }

    override fun toString(): String {
        return "BenchmarkModel(" +
                "options=$options,\n" +
                "timestamp=$timestamp,\n" +
                "keyboardApp=$keyboardApp,\n" +
                "correctChars=$correctChars,\n" +
                "correctWords=$correctWords,\n" +
                "errors=$errors,\n" +
                "charsPerSec=$charsPerSec,\n" +
                "wordsPerSec=$wordsPerSec,\n" +
                "totalWords=$totalWords,\n" +
                "timeElapsed=$timeElapsed,\n" +
                "backspace=$backspace,\n" +
                "keystrokes=$keystrokes,\n" +
                "characters=$characters,\n" +
                "keystrokesPerChar=$keystrokesPerChar,\n" +
                "minimumStringDistanceErrorRate=$minimumStringDistanceErrorRate,\n" +
                "inputString=$inputStream,\n" +
                "transcribedString=$transcribedString,\n" +
                ")"
    }


}